sklearn_numba_dpex/common/kernels.py

```diff
@@ -397,7 +429,149 @@ def partial_sum_reduction(
                 local_values[zero_idx] + local_values[one_idx]
             )
 
-    return work_group_size, partial_sum_reduction
+    if return_only_kernel:
+        return partial_sum_reduction
+
+    partial_sum_reductiom_nofunc = _make_partial_sum_reduction_2d_axis1_kernels(
+        n_rows, work_group_size, None, dtype, device, return_only_kernel=True
+    )
+
+    return work_group_size, (
+        (partial_sum_reduction, partial_sum_reductiom_nofunc),
+        reduction_block_size,
+    )
+
+
+@lru_cache
+def _make_partial_sum_reduction_2d_axis0_kernels(
+    n_cols,
+    work_group_size,
+    sub_group_size,
+    fused_unary_func,
+    dtype,
+    device,
+    return_only_kernel=False,
+):
+    zero = dtype(0.0)
+    one_idx = np.int64(1)
+    two_as_a_long = np.int64(2)
+
+    if fused_unary_func is None:
+
+        def fused_unary_func(x):
+            return x
+
+    fused_unary_func_ = dpex.func(fused_unary_func)
+
+    input_work_group_size = work_group_size
+    work_group_size = _check_max_work_group_size(
+        work_group_size, device, required_local_memory_per_item=np.dtype(dtype).itemsize
+    )
+    if work_group_size == input_work_group_size:
+        if (work_group_size % sub_group_size) != 0:
+            raise ValueError(
+                "Expected sub_group_size to divide work_group_size, but got "
+                f"sub_group_size={sub_group_size} and "
+                f"work_group_size={work_group_size}"
+            )
+        check_power_of_2(
+            n_sub_group_per_work_group := (work_group_size // sub_group_size)
+        )
+
+    else:
+        # Round work_group_size to the maximum smaller power-of-two multiple of
+        # `sub_group_size`
+        n_sub_group_per_work_group = 2 ** (
+            math.floor(math.log2(work_group_size / sub_group_size))
+        )
+        work_group_size = n_sub_group_per_work_group * sub_group_size
+
+    # Number of iteration in each execution of the kernel:
+    local_n_iterations = np.int64(math.floor(math.log2(n_sub_group_per_work_group)) - 1)
+
+    local_values_size = (n_sub_group_per_work_group, sub_group_size)
+    reduction_block_size = 2 * n_sub_group_per_work_group
+
+    # ???: how does this strategy compares to having each thread reducing N contiguous
+    # items ?
+    @dpex.kernel
+    # fmt: off
+    def partial_sum_reduction(
+        summands,    # IN        (n_rows, sum_axis_size)
+        result,      # OUT       (n_rows, math.ceil(size / (2 * reduction_block_size),)
+    ):
+
+        # fmt: on
+        # NB: This kernel only perform a partial reduction
+        group_id = dpex.get_group_id(zero_idx)
+        local_work_id = dpex.get_local_id(zero_idx)
+
+        sum_axis_size = summands.shape[zero_idx]
```

Comment:
I think I would rather rename `sum_axis_size` to `shape_0` or `summands_shape_0` if you prefer to be extra explicit.

Reply:
You're right the comment was wrong. It is fixed:

```        summands,    # IN        (sum_axis_size, n_cols)
        result,      # OUT       (math.ceil(size / (2 * reduction_block_size), n_cols)
```

Regarding naming for dimensions and sizes my idea was to use `size0`, `size1` as long as the number of dimension of the input is unknown (we can have size1 == None), and when it's known use `n_row`, `n_col` since it's 2d.